Scheduler notifications

CrazyTapok opened the pull request:
1. Screenshot: 
![image](https://github.com/user-attachments/assets/af730798-babe-4fd3-a2f3-48e487c4e55a)



2. What has been implemented: 
  - [x] Add EmailService
  - [x] Add NotificationSchedulerService
  - [x] Add NotificationController
  - [x] Add tests for EmailService, NotificationSchedulerService, NotificationController. 
  - [x] Fix other tests. 
  - [x] Add IEmailService and INotificationSchedulerService. 
  - [x] Add EmailNotification model. 
  - [x] Replace the usual constructor on primary. 
  - [x] Minor code fixes.

CrazyTapok committed:
feat: add scheduled notifications
Add EmailService. Add NotificationSchedulerService. Add NotificationController. Add tests for EmailService, NotificationSchedulerService, NotificationController. Fix other tests. Add IEmailService and INotificationSchedulerService. Add EmailNotification model. replace the usual constructor on primary. Minor code fixes.

CrazyTapok committed:
fix: the crown check of the expression in the model has been fixed.

svirzhevskiy commented on src/TaskManager.API/Contracts/Validators/EmailNotificationRequestValidator.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        RuleFor(request => request.EmailList)
+            .NotEmpty()
+            .Must(list => list.All(email => !string.IsNullOrWhiteSpace(email))).WithMessage("Email list contains invalid emails.");
```
мб юзать EmailAddress() вместо IsNullOrWhiteSpace?

svirzhevskiy commented on src/TaskManager.Core/Services/EmailService.cs:
```diff
@@ -0,0 +1,42 @@
+
+    public EmailService()
+    {
+        _smtpUser = Environment.GetEnvironmentVariable("SMTP_USER") ?? throw new ArgumentNullException(nameof(_smtpUser));
```
а можно это конфигурить как IOptions?

svirzhevskiy commented on src/TaskManager.API/Controllers/NotificationsController.cs:
```diff
@@ -0,0 +1,25 @@
+    [HttpPost("schedule")]
+    public IActionResult ScheduleEmailJob([FromBody] EmailNotificationRequest request)
+    {
+        var emailNotification = request.ToEmailNotification();
```
то есть чел задает какое-то body и время и одно и то же будет отсылаться всем жертвам постоянно?)
Это может и работает, но выглядит бесполезным :)
Я представлял что-то как дайджест апдейтов или типо того. Например, чтобы манагер каждое утро получал список тасок с их статусами со своих проектов

CrazyTapok committed:
feat: the functionality of scheduled notifications has been redesigned
Fix EmailService. Fix NotificationSchedulerService. Fix NotificationController. Fix tests for EmailService, NotificationSchedulerService, NotificationController. Fix IEmailService and INotificationSchedulerService. Fix EmailNotification model. replace the usual constructor on primary. Minor code fixes. Add GenerateProjectNotificationKey, SmtpSettings model,  NotificationSettings model. Add ISmtpClient. Add SmtpClientWrapper.

CrazyTapok committed:
fix: fix docker-compose
The environment section has been updated.

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/Configuration/NotificationSettings.cs:
```diff
@@ -0,0 +1,7 @@
+
+public class NotificationSettings
+{
+    public int Hour { get; set; }
+    public int Minute { get; set; }
```
почему решил так время передавать? Потенциально даешь пространство для ошибки. Мб лучше использовать тип для времени?

svirzhevskiy commented on src/TaskManager.Core/Services/EmailService.cs:
```diff
@@ -0,0 +1,36 @@
+        message.Body = builder.ToMessageBody();
+
+        await _smtpClient.SendAsync(message, cancellationToken);
+        await _smtpClient.DisconnectAsync(true, cancellationToken);
```
предложение по улучшению читабельности: если передаешь в метод литерал, то подписывать его именованным параметром. Я щас даже догадки не имею, что значит этот true :)

svirzhevskiy commented on src/TaskManager.API/Controllers/NotificationController.cs:
```diff
@@ -0,0 +1,34 @@
+    [HttpPost("enable")]
+    public IActionResult EnableNotifications(Guid projectId, [FromBody] EmailNotificationRequest request)
+    {
+        var emailNotification = request.ToEmailNotification();
```
здесь получается можно любой имейл указать для нотификаций? не думал указывать просто менеджера айди, чтобы отсылать можно было только зареганым пользователям? 
Тут правильного варианта нет, просто спорный вопрос. Мне кажется, что когда ты выбираешь получать уведомления и у тебя уже есть пользователь, то другой имейл выбрать нельзя

svirzhevskiy commented on src/TaskManager.API/Controllers/NotificationController.cs:
```diff
@@ -0,0 +1,34 @@
+    private readonly IProjectNotificationService _projectNotificationService = projectNotificationService;
+
+    [Authorize(Roles = "Admin,ProjectManager")]
+    [HttpPost("enable")]
```
получилось не особо по rest. Должен быть объект в урле, а действия как-то через тело запроса, в параметре,  захендлить. 
Но настаивать переделывать не буду, потому как никто не живет по тотальному rest) Чисто, подсветил момент

svirzhevskiy commented on src/TaskManager.API/Program.cs:
```diff
@@ -41,6 +43,8 @@
 CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("en-US");
 
 builder.Services.Configure<JwtSettings>(builder.Configuration.GetSection("JWT"));
+builder.Services.Configure<SmtpSettings>(builder.Configuration.GetSection("SMTP"));
```
можно указывать названия секций в статическом поле внутри класса сетингов или использовать nameof(TSettings) и называть сеттинги так же как класс. А то вариант с просто литерала в коде не оч. Лучше держать все в константах каких-то

svirzhevskiy commented on src/TaskManager.Core/Services/ProjectReportService.cs:
```diff
@@ -0,0 +1,34 @@
+
+    public async Task<string> GenerateProjectReportAsync(Guid projectId, CancellationToken cancellationToken = default)
+    {
+        var tasks = await _taskService.GetTasksByProjectIdAsync(projectId, cancellationToken);
```
а знаешь про AsNoTracking() в ef? Я щас подумал, может его везде добавить в репозитории, кроме GetById? Потому что апдейтить скорее всего только с ним будешь

svirzhevskiy replied:
пропустил коммент?

svirzhevskiy commented on src/TaskManager.Core/Services/ProjectNotificationService.cs:
```diff
@@ -0,0 +1,40 @@
+
+    public void SendProjectReport(Guid projectId, EmailNotification emailNotification, CancellationToken cancellationToken = default)
+    {
+        var report = projectReportService.GenerateProjectReportAsync(projectId, cancellationToken);
```
а где этот репорт потом юзается?

svirzhevskiy replied:
он должен попасть в emailNotification? Если это не поймали юнит тесты, значит херово тесты написал)))

svirzhevskiy commented on src/TaskManager.Core/Services/ProjectNotificationService.cs:
```diff
@@ -0,0 +1,40 @@
+    public void SendProjectReport(Guid projectId, EmailNotification emailNotification, CancellationToken cancellationToken = default)
+    {
+        var report = projectReportService.GenerateProjectReportAsync(projectId, cancellationToken);
+        emailService.SendEmailAsync(projectId, emailNotification, cancellationToken);
```
тут специально не await ишь? если это в джобе крутится мб стоит await тить все равно? 
Без await, например, эксепшн теряется

svirzhevskiy replied:
мб и не надо await ить, это вопрос

svirzhevskiy commented on src/TaskManager.Core/Services/ProjectReportService.cs:
```diff
@@ -0,0 +1,34 @@
+
+internal class ProjectReportService(ITaskService taskService) : IProjectReportService
+{
+    private readonly ITaskService _taskService = taskService ?? throw new ArgumentNullException(nameof(taskService));
```
разве эксепшн не упадет при попытке создать ProjectReportService, если taskService не будет? Есть подозрение, что этот throw недостижим

svirzhevskiy commented on src/TaskManager.Core/Services/ProjectNotificationService.cs:
```diff
@@ -0,0 +1,40 @@
+
+    public void EnableNotifications(Guid projectId, EmailNotification emailNotification, CancellationToken cancellationToken = default)
+    {
+        var jobKey = JobKeyGenerator.GenerateProjectNotificationKey(projectId);
+
+        recurringJobManager.AddOrUpdate(
+            jobKey,
+            () => SendProjectReport(projectId, emailNotification, cancellationToken),
+            Cron.Daily(_settings.Hour, _settings.Minute)
+        );
```
я щас думаю... Правильно понимаю, что после перезапуска приложения, все джобы исчезнут и нужно будет создавать заново?

svirzhevskiy replied:
Если да, то нужен другой подход:
при старте приложения запускается джоба, которая крутится все время жизни приложения. Она читает из базы какие нужно отправить нотификации и отправляет их. А через АПИ мы указываем кому какие нотификации отправлять и сохраняем в базу для джобы

CrazyTapok committed:
feat: redesigned the functionality for daily mailing
add a service for collecting project data with mailing, add new property in Employee model, add service to create jobSheduler, add minor changes, fix tests

CrazyTapok committed:
fix: minor changes

svirzhevskiy commented on src/TaskManager.Core/Services/DailyNewsletterSchedulerService.cs:
```diff
@@ -0,0 +1,33 @@
+        IJobSchedulerWrapper jobSchedulerWrapper)
+    {
+        _newsletterReportService = newsletterJobService ?? throw new ArgumentNullException(nameof(newsletterJobService));
+        _notificationSettings = notificationSettings?.Value ?? throw new ArgumentNullException(nameof(notificationSettings));
+        _jobSchedulerWrapper = jobSchedulerWrapper ?? throw new ArgumentNullException(nameof(jobSchedulerWrapper));
```
писал коммент про такое уже. 
С сеттингами скорее всего имеет смысл проверка, но с сервисом разве ошибка падает не в di?

CrazyTapok replied:
Ты прав, проверка notificationSettings на null оправдана, т.к. IOptions<T> используется для обертки настроек, и Value может отсутствовать. А _newsletterReportService, регистрируется в контейнере зависимостей (DI), и если его не удается разрешить, то ошибка произойдет еще при создании экземпляра DailyNewsletterSchedulerService.

svirzhevskiy commented on src/TaskManager.Core/Services/DailyNewsletterSchedulerService.cs:
```diff
@@ -0,0 +1,33 @@
+        var minutes = _notificationSettings.NotificationTime.Minutes;
+
+        _jobSchedulerWrapper.ScheduleJob(
+            jobId: "daily-newsletter-job",
```
мб тоже вынести в конфиг, он же есть у тебя

CrazyTapok replied:
да, ты прав, упустил

svirzhevskiy commented on src/TaskManager.Core/Services/NewsletterReportService.cs:
```diff
@@ -0,0 +1,50 @@
+
+        foreach (var employee in employees)
+        {
+            var projects = await _projectService.GetProjectsByEmployeeIdAsync(employee.Id, cancellationToken);
```
нихуево ты базу подолбишь такими запросами в цикле, если много будет employee :)

svirzhevskiy commented on src/TaskManager.Core/Services/SmtpClientWrapper.cs:
```diff
@@ -0,0 +1,33 @@
+
+    public async Task ConnectAsync(string host, int port, SecureSocketOptions options, CancellationToken cancellationToken = default)
+    {
+        await _smtpClient.ConnectAsync(host, port, options, cancellationToken);
```
можно тут не делать async await? чтобы сэкономить на разворачивании стейт машины здесь

CrazyTapok replied:
ты имеешь ввиду именно в этом методе или во всех? 
да, можно

svirzhevskiy commented on src/TaskManager.Core/Services/SmtpClientWrapper.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using MailKit.Net.Smtp;
+using MailKit.Security;
+using MimeKit;
```
что то мне не нравится, что у тебя в кор пошла зависимость от стороней либы. По аналогии с репозиториями, тут будет интерфейс, а реализация отправки сообщений - это инфраструктура. И с тестами тогда стакнется

svirzhevskiy replied:
я так понял, это тянется сюда из-за моделек. Добавь свои модели с инфой, которая важна только для кора

svirzhevskiy commented on src/TaskManager.Core/Services/JobSchedulerWrapper.cs:
```diff
@@ -0,0 +1,18 @@
+
+internal class JobSchedulerWrapper : IJobSchedulerWrapper
+{
+    private readonly IRecurringJobManager _recurringJobManager;
```
и зависимость на хенфаер в коре тоже не очень красиво выглядит. В идеале, твой кор не зависит ни от чего, а все зависит от него.
![image](https://github.com/user-attachments/assets/bade1085-f94c-4dcc-9845-2e921c56f7a7)
* тут core - это domain + application

CrazyTapok committed:
fix: correction after the review
the project structure chang, chang made to comply with the principles of clean architecture, optimiz the command for generating the report, add the use of AsNoTracking() to the repository

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/Configuration/NotificationSettings.cs:
```diff
@@ -0,0 +1,11 @@
+    public static readonly string SectionName = "NOTIFICATION";
+    public required string Id { get; set; }
+    public TimeSpan Time { get; set; }
+    public string CronExpression => Cron.Daily(Time.Hours, Time.Minutes);
```
это тоже можно в конфиг

CrazyTapok committed:
feat: delete time in NotificationSettings model, use CronExpression

CrazyTapok committed:
fix: fix docker-compose

svirzhevskiy commented on src/TaskManager.Core/Services/Scheduling/NewsletterReportService.cs:
```diff
@@ -0,0 +1,47 @@
+                    Body = reportContent
+                };
+
+                await _emailService.SendEmailAsync(emailNotification, cancellationToken);
```
здесь ведь имейлы не зависят друг от друга. Можно все таски в параллель запустить

svirzhevskiy commented on src/TaskManager.Core/Interfaces/Services/Email/ISmtpClientWrapper.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using MailKit.Security;
+using MimeKit;
```
все равно затянул зависимость от конкретной реализации в кор. Тебе нужно создать модель с параметрами для smtpClient в коре и использовать ее. А в конкрентой реализации мапить параметры из этой модели в модель для конкретно этой реализации.

svirzhevskiy replied:
открывай кароч core.csprj и смотри что ты туда затянул и выпиливай все лишнее

svirzhevskiy commented on src/TaskManager.Core/TaskManager.Core.csproj:
```diff
@@ -8,9 +8,12 @@
 
   <ItemGroup>
     <PackageReference Include="BCrypt.Net-Next" Version="4.0.3" />
+    <PackageReference Include="Hangfire.Core" Version="1.8.18" />
+    <PackageReference Include="MailKit" Version="4.11.0" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="9.0.0" />
     <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="9.0.0" />
     <PackageReference Include="Microsoft.Extensions.Options" Version="9.0.0" />
+    <PackageReference Include="MimeKit" Version="4.11.0" />
```
в это три новых строки точно убераем, возможно что-то еще

svirzhevskiy reviewed (commented):
нужно дальше отвязывать кор от внешних зависимостей

svirzhevskiy commented on src/TaskManager.Core/Services/Scheduling/NewsletterReportService.cs:
```diff
@@ -0,0 +1,47 @@
+        _emailService = emailService;
+    }
+
+    public async Task ExecuteDailyJob(CancellationToken cancellationToken = default)
```
я бы назвал более бизнесово, а не execute job

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -55,6 +55,6 @@ public async Task DeleteAsync(Guid id, CancellationToken cancellationToken = def
 
     public virtual Task<List<TModel>> FindAsync(Expression<Func<TModel, bool>> predicate, CancellationToken cancellationToken = default)
```
сука, гитхаб не дает оставить коммент не по теме мра)))
Уверен, что на 45ой строке у тебя канселейшн токен передается как канселейшн токен, а не как параметр для поиска? Потому что там параметры передуются через params

svirzhevskiy replied:
аналогично на 22ой

svirzhevskiy replied:
и еще вопрос в тему удаления:
как работает ExecuteDeleteAsync, может он позволит удалить в один запрос к бд?

CrazyTapok replied:
Да, ExecuteDeleteAsync как раз для этого и придумали в EF Core 7 — чтобы удалять записи напрямую в базе одним SQL‑запросом, без загрузки сущностей в память и без работы change tracker’а.

Если тебе нужен soft delete в один запрос, можно использовать `ExecuteUpdateAsync`:
```
await _dbSet
    .Where(e => e.Id == id)
    .ExecuteUpdateAsync(s => s.SetProperty(e => e.IsDeleted, true), cancellationToken);
```

svirzhevskiy replied:
почему не заюзал тогда?)

svirzhevskiy commented on src/TaskManager.Infrastructure/Services/Reporting/HtmlProjectReportService.cs:
```diff
@@ -0,0 +1,33 @@
+
+internal class HtmlProjectReportService : IProjectReportService
+{
+    public string GenerateProjectReport(Guid projectId, List<Task> tasks)
```
спроси у нейросетки, она считает, что это в инфраструктуру или кор? тока подробно ей опиши ситуацию) мне тож интересно)

CrazyTapok replied:
Что делает HtmlProjectReportService

1. **Входные данные** — `Guid projectId` и `List<Task> tasks` (с доменной моделью `Task`).
2. **Выходные данные** — готовая HTML-строка.
3. **Логика** — полностью завязана на формат представления (HTML), а не на бизнес-правила.
4. **Зависимости** — `StringBuilder`, HTML-разметка, форматирование дат.

Почему это не Core

- **Core** должен уметь сказать: "Сгенерируй отчёт по проекту" — но не знать, что это HTML. В Core можно оставить интерфейс `IProjectReportService` и, например, модель `ProjectReport` в виде чистых данных.
- HTML — это уже конкретная технология представления, а значит, это инфраструктурная деталь.
- Если завтра захочешь PDF или Markdown, Core не должен меняться — только реализация в Infrastructure.

Вывод
`HtmlProjectReportService` — однозначно в Infrastructure, потому что:

- Он жёстко привязан к HTML-формату.
- Не является бизнес-правилом, а лишь способом представления данных.
- Его можно заменить на другую реализацию, не трогая Core.

svirzhevskiy commented on src/TaskManager.API/Controllers/TaskController.cs:
```diff
@@ -3,7 +3,7 @@
+using TaskManager.Core.Interfaces.Services.ProjectManagement;
 
 namespace TaskManager.API.Controllers;
 
```
у теяб тут в 21ой строке не может быть тull, потому что ты кидаешь ошибку внутри сервиса

svirzhevskiy replied:
у тебя во всех контроллерах замиксовано два подхода по инициализации: ты сервисы определяешь как отдельные проперти, а вместо конструктора у тебя в скобочки инжектится
<img width="1003" height="168" alt="image" src="https://github.com/user-attachments/assets/c544c992-9620-4f54-b949-51d92ea9cdf9" />
Я бы выбрал какой-то один подход

svirzhevskiy commented on src/TaskManager.API/Controllers/TaskController.cs:
```diff
@@ -3,7 +3,7 @@
+using TaskManager.Core.Interfaces.Services.ProjectManagement;
 
 namespace TaskManager.API.Controllers;
 
```
на 41ой строке CreatedAtAction это не про вьюхи? чем отличается от Created?

CrazyTapok replied:
Не, CreatedAtAction не имеет отношения к вьюхам — это чисто про HTTP-ответ и REST‑подход.

- Возвращает HTTP 201 Created.

- Добавляет в заголовок ответа Location ссылку на только что созданный ресурс.

- Ссылка строится на основе указанного метода контроллера (nameof(GetByIdAsync)) и переданных route‑параметров (new { id = response.Id }).

- Тело ответа (response) — это сам созданный объект.

`return CreatedAtAction(nameof(GetByIdAsync), new { id = response.Id }, response);`

Это означает:

> "Вот тебе созданный объект, и если хочешь получить его позже — иди по этому маршруту GetByIdAsync с id = response.Id."

Почему CreatedAtAction лучше в API Это REST‑подход: клиент сразу получает не только данные, но и точный адрес, по которому можно запросить этот же ресурс.

CrazyTapok committed:
fix: correction after the review
Divided the logic in the project according to the principles of clean architecture. Removed external dependencies from Core. Renamed ExecuteDailyJob to SendDailyProjectReportsAsync. I used a unified approach for initialization in all controllers. Fix tests.

svirzhevskiy commented:
вроде все пофиксил, остался только момент с удалением. Мне кажется, лучше в один запрос удалять

CrazyTapok committed:
fix: correction after the review
Change method DeleteAsync in Repository, change test for method
